internal/gengapic/gengapic.go

```diff
@@ -95,15 +97,53 @@ func Gen(genReq *plugin.CodeGeneratorRequest) (*plugin.CodeGeneratorResponse, er
 	}
 	g.metadata.LibraryPackage = g.opts.pkgPath
 
+	if !g.opts.omitSnippets {
+		// Initialize the model that will collect snippet metadata.
+		sm, err := snippets.NewMetadata(protoPkg, g.metadata.LibraryPackage, g.opts.pkgName)
+		if err != nil {
+			return &g.resp, err
+		}
+		g.snippetMetadata = sm
+	}
+
 	for _, s := range genServs {
 		// TODO(pongad): gapic-generator does not remove the package name here,
 		// so even though the client for LoggingServiceV2 is just "Client"
 		// the file name is "logging_client.go".
 		// Keep the current behavior for now, but we could revisit this later.
-		outFile := pbinfo.ReduceServName(s.GetName(), "")
-		outFile = camelToSnake(outFile)
+		servName := pbinfo.ReduceServName(s.GetName(), "")
+		outFile := camelToSnake(servName)
 		outFile = filepath.Join(g.opts.outDir, outFile)
 
+		if !g.opts.omitSnippets {
```

Comment:
Not saying we should do it now. But I think if we ever refactor in the future we should avoid all of these if checks in the main generation methods. I think it would be cleaner to call methods and just have them do nothing if flags are not set. I think it would make the code more readable. I know there is presidence for this pattern rn though. But something we should consider should we do a rewrite of this package.

Reply:
I'm open to attempting this refactoring now to see how it looks. The new method would probably need a quite a few parameters, but it is probably worth it. I agree with this suggestion.

Reply:
I think we should land this first regardless as this is already a large PR

Reply:
I pushed a commit moving this logic into a new file `internal/gengapic/snippets.go`, which I think cleans up the existing files quite a bit, at the cost of a bit of indirection. It may also make it possible to improve the test coverage.